File: simulators/c++2/src_simulator/TMLTask.cpp
/*Copyright or (C) or Copr. GET / ENST, Telecom-Paris, Daniel Knorreck,
Ludovic Apvrille, Renaud Pacalet
 *
 * ludovic.apvrille AT telecom-paristech.fr
 *
 * This software is a computer program whose purpose is to allow the
 * edition of TURTLE analysis, design and deployment diagrams, to
 * allow the generation of RT-LOTOS or Java code from this diagram,
 * and at last to allow the analysis of formal validation traces
 * obtained from external tools, e.g. RTL from LAAS-CNRS and CADP
 * from INRIA Rhone-Alpes.
 *
 * This software is governed by the CeCILL  license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 *
 */

#include <TMLTask.h>
#include <TMLCommand.h>
#include <CPU.h>

TMLTask::TMLTask(unsigned int iPriority, std::string iName, CPU* iCPU):_name(iName), _priority(iPriority), _endLastTransaction(0), _currCommand(0), _cpu(iCPU), _previousTransEndTime(0), _comment(0), _busyCycles(0), _CPUContentionDelay(0), _noCPUTransactions(0) {
	_myid=++_id;
	_cpu->registerTask(this);
#ifdef ADD_COMMENTS
	_commentList.reserve(BLOCK_SIZE);
#endif
	_transactList.reserve(BLOCK_SIZE);
}

TMLTask::~TMLTask(){
	CommentList::iterator i;
#ifdef ADD_COMMENTS
	for(i=_commentList.begin(); i != _commentList.end(); ++i){
		delete *i;
	}
#endif
	if (_comment!=0) delete [] _comment;
}

unsigned int TMLTask::getPriority() const{
	return _priority;
}

TMLTime TMLTask::getEndLastTransaction() const{
	return _endLastTransaction;
}

TMLCommand* TMLTask::getCurrCommand() const{
	return _currCommand;
}

void TMLTask::setCurrCommand(TMLCommand* iCurrCommand){
	//std::cout << _name << "currentcommand: " << iCurrCommand->toString() << std::endl;
	_currCommand=iCurrCommand;
}

CPU* TMLTask::getCPU(){
	return _cpu;
}

std::string TMLTask::toString(){
	return _name;
}

std::string TMLTask::toShortString(){
	std::ostringstream outp;
	outp << "ta" << _myid;
	return outp.str();
}

unsigned int TMLTask::getID(){
	return _myid;
}

#ifdef ADD_COMMENTS
void TMLTask::addComment(Comment* iComment){
	_commentList.push_back(iComment);
}

std::string TMLTask::getNextComment(bool iInit, Comment*& oComment){
	if (iInit) _posCommentList=_commentList.begin();
	if (_posCommentList == _commentList.end()){
		//std::cout << "ret0\n";
		oComment=0;
		return std::string("no more comment");
	}
	//std::cout << "NON0\n";
	oComment=*_posCommentList;
	_posCommentList++;
	return ((oComment->_command==0)?_comment[oComment->_actionCode]:oComment->_command->getCommentString(oComment));
}
#endif

void TMLTask::addTransaction(TMLTransaction* iTrans){
	_transactList.push_back(iTrans);
	_endLastTransaction=iTrans->getEndTime();
	_busyCycles+=iTrans->getOperationLength();
	if(iTrans->getChannel()==0){
		_noCPUTransactions++;
		_CPUContentionDelay+=iTrans->getStartTime()-iTrans->getRunnableTime();
	}
}

TMLTime TMLTask::getNextSignalChange(bool iInit, std::string& oSigChange, bool& oNoMoreTrans){
	std::ostringstream outp;
	if (iInit){
		//std::cout << "Init" << std::endl; 
		_posTrasactList=_transactList.begin();
		//std::cout << "Init2" << std::endl; 
		_previousTransEndTime=0;
		_vcdOutputState=END_TRANS;
	}
	if (_posTrasactList == _transactList.end()){
		if (iInit || _transactList.back()->getTerminatedFlag()){
			outp << "r" << TERMINATED << " ta" << _myid;
		}else{
			outp << "r" << SUSPENDED << " ta" << _myid;
		}
		oSigChange=outp.str();
		oNoMoreTrans=true;
		return _previousTransEndTime;
	}else{
		//std::cout << "VCD out trans: " << (*_posTrasactList)->toShortString() << std::endl;
		TMLTransaction* aCurrTrans=*_posTrasactList;
		oNoMoreTrans=false;
		switch (_vcdOutputState){
			case END_TRANS:
				if (aCurrTrans->getRunnableTime()==_previousTransEndTime){
					outp << "r" << RUNNABLE << " ta" << _myid;
					_vcdOutputState=START_TRANS;
				}else{
					outp << "r" << SUSPENDED << " ta" << _myid;
					if (aCurrTrans->getRunnableTime()==aCurrTrans->getStartTimeOperation()){
						_vcdOutputState=START_TRANS;
					}else{
						_vcdOutputState=BETWEEN_TRANS;
					}
				}
				oSigChange=outp.str();
				return _previousTransEndTime;
			break;
			case BETWEEN_TRANS:
				outp << "r" << RUNNABLE << " ta" << _myid;
				oSigChange=outp.str();
				_vcdOutputState=START_TRANS;
				return aCurrTrans->getRunnableTime();
			break;
			case START_TRANS:
				outp << "r" << RUNNING << " ta" << _myid;
				oSigChange=outp.str();
				do{
					_previousTransEndTime=(*_posTrasactList)->getEndTime();
					_posTrasactList++;
				}while (_posTrasactList != _transactList.end() && (*_posTrasactList)->getStartTimeOperation()==_previousTransEndTime);
				_vcdOutputState=END_TRANS;
				return aCurrTrans->getStartTimeOperation();
			break;
		}
	}
}

/*unsigned int TMLTask::getBusyCycles(){
	return 0;
}*/

std::ostream& TMLTask::writeObject(std::ostream& s){
	unsigned int aCurrCmd;
	WRITE_STREAM(s,_endLastTransaction);
	if (_currCommand==0){
		aCurrCmd=0;
		WRITE_STREAM(s,aCurrCmd);
	}else{
		aCurrCmd=(unsigned int)_currCommand-(unsigned int)this;
		WRITE_STREAM(s,aCurrCmd);
		_currCommand->writeObject(s);
	}
	return s;
}

std::istream& TMLTask::readObject(std::istream& s){
	unsigned int aCurrCmd;
	READ_STREAM(s, _endLastTransaction);
	READ_STREAM(s, aCurrCmd);
	_currCommand=(aCurrCmd==0)?0:(TMLCommand*)(aCurrCmd+((unsigned int)this));
	if (_currCommand!=0) _currCommand->readObject(s);
	return s;
}

void TMLTask::streamBenchmarks(std::ostream& s){
	s << "*** Task " << _name << " ***\n"; 
	s << "Execution time: " << _busyCycles << std::endl;
	if (_noCPUTransactions!=0) s << "Average CPU contention delay: " << ((float)_CPUContentionDelay)/((float)_noCPUTransactions) << std::endl;
}

File: src/main/java/tmltranslator/TMLTask.java
/* Copyright or (C) or Copr. GET / ENST, Telecom-Paris, Ludovic Apvrille
 * 
 * ludovic.apvrille AT enst.fr
 * 
 * This software is a computer program whose purpose is to allow the
 * edition of TURTLE analysis, design and deployment diagrams, to
 * allow the generation of RT-LOTOS or Java code from this diagram,
 * and at last to allow the analysis of formal validation traces
 * obtained from external tools, e.g. RTL from LAAS-CNRS and CADP
 * from INRIA Rhone-Alpes.
 * 
 * This software is governed by the CeCILL  license under French law and
 * abiding by the rules of distribution of free software.  You can  use,
 * modify and/ or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * 
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 * 
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 * 
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 */


package tmltranslator;

import myutil.TraceManager;

import java.util.*;

/**
 * Class TMLTask
 * Creation: 17/11/2005
 *
 * @author Ludovic APVRILLE
 * @version 1.0 17/11/2005
 */
public class TMLTask extends TMLElement {
    protected TMLActivity activity;
    private boolean isRequested = false;
    private TMLRequest request;
    private List<TMLAttribute> attributes;
    private boolean mustExit = false;
    private int priority;
    private Set<TMLChannel> channelsList;
    private Set<TMLChannel> readTMLChannelsList;
    private Set<TMLChannel> writeTMLChannelsList;
    private Set<TMLEvent> eventsList;
    private int operationType;
    private String operation = "";
    private boolean isDaemon;
    private String operationMEC;
    private boolean isAttacker;

    public TMLTask(String name, Object referenceToClass, Object referenceToActivityDiagram) {
        super(name, referenceToClass);
        //TraceManager.addDev("Creating new TMLTask:" + name);
        activity = new TMLActivity(name + "activity_diagram", referenceToActivityDiagram);
        attributes = new ArrayList<TMLAttribute>();
        channelsList = new HashSet<TMLChannel>();
        readTMLChannelsList = new HashSet<TMLChannel>();
        writeTMLChannelsList = new HashSet<TMLChannel>();
        eventsList = new HashSet<TMLEvent>();
    }

    public void setRequested(boolean _b) {
        isRequested = _b;
    }

    public void setDaemon(boolean _b) {
        isDaemon = _b;
    }

    public boolean isDaemon() {return isDaemon;}

    public void setPriority(int _priority) {
        priority = _priority;
    }

    public int getPriority() {
        return priority;
    }

    public boolean isRequested() {
        return isRequested;
    }

    public boolean isAttacker() {
        return isAttacker;
    }

    public void setAttacker(boolean a) {
        isAttacker = a;
    }

    public void setRequest(TMLRequest _request) {
        request = _request;
    }

    public TMLRequest getRequest() {
        return request;
    }

    public void addAttribute(TMLAttribute _tmla) {
        attributes.add(_tmla);
    }


    public void addAttributeIfApplicable(TMLAttribute _tmla) {
        for(TMLAttribute att: attributes) {
            if (att.getName().compareTo(_tmla.getName()) == 0) {
                return;
            }
        }
        attributes.add(_tmla);
    }

    public List<TMLAttribute> getAttributes() {
        return attributes;
    }

    public boolean hasCommand(int commandID) {
        TMLActivityElement tmlae;
        for (int i = 0; i < activity.nElements(); i++) {
            tmlae = activity.get(i);
            if (tmlae.getID() == commandID) {
                return true;
            }
        }
        return false;
    }

    public TMLActivityElement getElementByID(int commandID) {
        TMLActivityElement tmlae;
        for (int i = 0; i < activity.nElements(); i++) {
            tmlae = activity.get(i);
            if (tmlae.getID() == commandID) {
                return tmlae;
            }
        }
        return null;
    }

    public String[] makeCommandIDs() {
        String[] list = new String[activity.nElements()];
        TMLActivityElement tmlae;
        for (int i = 0; i < activity.nElements(); i++) {
            tmlae = activity.get(i);
            list[i] = tmlae.getName() + " (" + tmlae.getID() + ")";
        }
        return list;
    }

    public String[] makeVariableIDs() {
        String[] list = new String[attributes.size()];
        int cpt = 0;

        for (TMLAttribute att : attributes) {
            list[cpt] = att.getName() + " (" + att.getID() + ")";
            cpt++;
        }
        return list;
    }

    public TMLAttribute getAttributeByName(String _name) {
        for (TMLAttribute attribute : attributes) {
            if (attribute.getName().compareTo(_name) == 0) {
                return attribute;
            }
        }
        return null;
    }

    //For SDR operations only 1 channel
    public List<TMLReadChannel> getReadChannels() {

        List<TMLReadChannel> list = new ArrayList<TMLReadChannel>();
        for (int i = 0; i < getActivityDiagram().nElements(); i++) {
            if (getActivityDiagram().get(i) instanceof TMLReadChannel) {
                list.add((TMLReadChannel) getActivityDiagram().get(i));
                //TraceManager.addDev( "Element: " + task.getActivityDiagram().get(i).toString() );
            }
        }

        return list;
    }

    //For SDR operations, only 1 channel
    public List<TMLWriteChannel> getWriteChannels() {

        List<TMLWriteChannel> list = new ArrayList<TMLWriteChannel>();
        for (int i = 0; i < getActivityDiagram().nElements(); i++) {
            if (getActivityDiagram().get(i) instanceof TMLWriteChannel) {
                list.add((TMLWriteChannel) getActivityDiagram().get(i));
                //TraceManager.addDev( "Element: " + task.getActivityDiagram().get(i).toString() );
            }
        }
        return list;
    }

    public List<TMLSendEvent> getSendEvents() {

        List<TMLSendEvent> list = new ArrayList<TMLSendEvent>();

        for (int i = 0; i < getActivityDiagram().nElements(); i++) {
            if (getActivityDiagram().get(i) instanceof TMLSendEvent) {
                list.add((TMLSendEvent) getActivityDiagram().get(i));
            }
        }

        return list;
    }

    public List<TMLWaitEvent> getWaitEvents() {
        List<TMLWaitEvent> list = new ArrayList<TMLWaitEvent>();

        for (int i = 0; i < getActivityDiagram().nElements(); i++) {
            if (getActivityDiagram().get(i) instanceof TMLWaitEvent) {
                list.add((TMLWaitEvent) getActivityDiagram().get(i));
            }
        }

        return list;
    }

    public String getID0() {

        if (getReadChannels().size() > 0) {
            return getReadChannels().get(0).toString().split("__")[1];
        } else {
            return "";
        }
    }

    public String getOD0() {

        if (getWriteChannels().size() > 0) {
            return getWriteChannels().get(0).toString().split("__")[1];
        } else {
            return "";
        }
    }

    public String getTaskName() {
        if (getName().indexOf("__") == -1) {
            return getName();
        }
        return getName().split("__")[1];
    }


    public TMLActivity getActivityDiagram() {
        return activity;
    }

    public void addElement(TMLActivityElement prev, TMLActivityElement succ) {
        if (activity == null) {
            return;
        }
        activity.addElement(succ);
        prev.addNext(succ);
    }


    public void setExit(boolean b) {
        mustExit = b;
    }

    public boolean exits() {
        return mustExit;
    }

    public boolean has(TMLActivityElement _elt) {
        if (activity == null) {
            return false;
        }

        return activity.contains(_elt);
    }

    public String getNameExtension() {
        return "task__";
    }

    public int getMaximumSelectEvtSize() {
        return activity.getMaximumSelectEvtSize();
    }

    public String getAttributeString() {
        String ret = "";
        for (TMLAttribute attribute : attributes) {
            ret += attribute.toString() + " / ";
        }
        return ret;
    }

    public boolean hasTMLRandom() {
        TMLActivityElement element;
        for (int i = 0; i < activity.nElements(); i++) {
            element = activity.get(i);
            if (element instanceof TMLRandom) {
                return true;
            }
        }
        return false;
    }

    public void removeAllRandomSequences() {
        activity.removeAllRandomSequences(this);
    }

    public Vector<String> getAllAttributesStartingWith(String _name) {
        Vector<String> v = new Vector<String>();
        for (TMLAttribute attribute : attributes) {
            if (attribute.getName().startsWith(_name)) {
                v.add(attribute.getName());
            }
        }
        return v;
    }

    public Vector<TMLAttribute> getAllTMLAttributesStartingWith(String _name) {
        Vector<TMLAttribute> v = new Vector<TMLAttribute>();
        for (TMLAttribute attribute : attributes) {
            if (attribute.getName().startsWith(_name)) {
                v.add(attribute);
            }
        }
        return v;
    }

    public int computeMaxID() {
        int max = getID();
        if (activity != null) {
            max = Math.max(max, activity.computeMaxID());
        }
        return max;
    }

    public void computeCorrespondance(TMLElement[] _correspondance) {
        _correspondance[getID()] = this;
        if (activity != null) {
            activity.computeCorrespondance(_correspondance);
        }

    }

    public void replaceWaitEventWith(TMLEvent oldEvt, TMLEvent newEvt) {
        activity.replaceWaitEventWith(oldEvt, newEvt);
    }

    public void replaceSendEventWith(TMLEvent oldEvt, TMLEvent newEvt) {
        activity.replaceSendEventWith(oldEvt, newEvt);
    }

    public void replaceReadChannelWith(TMLChannel oldChan, TMLChannel newChan) {
        activity.replaceReadChannelWith(oldChan, newChan);
    }

    public void replaceWriteChannelWith(TMLChannel oldChan, TMLChannel newChan) {
        activity.replaceWriteChannelWith(oldChan, newChan);
    }

    public void addSendEventAfterWriteIn(TMLChannel chan, TMLEvent evt, String action) {
        activity.addSendEventAfterWriteIn(chan, evt, action);
    }

    public void addSendAndReceiveEventAfterWriteIn(TMLChannel chan, TMLEvent evt1, TMLEvent evt2, String action1, String action2) {
        activity.addSendAndReceiveEventAfterWriteIn(chan, evt1, evt2, action1, action2);
    }

    public void addTMLChannel(TMLChannel _ch) {
        channelsList.add(_ch);
    }

    public void addReadTMLChannel(TMLChannel _ch) {
        readTMLChannelsList.add(_ch);
    }

    public void addWriteTMLChannel(TMLChannel _ch) {
        writeTMLChannelsList.add(_ch);
    }

    /*public List<TMLChannel> getTMLChannels() {
        return new ArrayList<TMLChannel>(channelsList);
    }*/

    public List<TMLChannel> getReadTMLChannels() {
        return new ArrayList<TMLChannel>(readTMLChannelsList);
    }

    public List<TMLChannel> getWriteTMLChannels() {
        return new ArrayList<TMLChannel>(writeTMLChannelsList);
    }

    public void addTMLEvent(TMLEvent _evt) {
        eventsList.add(_evt);
    }

    public List<TMLEvent> getTMLEvents() {
        return new ArrayList<TMLEvent>(eventsList);
    }

    public void addOperationType(int _operationType) {
        operationType = _operationType;
    }

    public int getOperationType() {
        return operationType;
    }

    public void addOperationMEC(String _operation) {
        operationMEC = _operation;
    }

    public String getOperationMEC() {
        return operationMEC;
    }

    public void addOperation(String _operation) {
        operation = _operation;
    }

    public String getOperation() {
        return operation;
    }

    public void removeEmptyInfiniteLoop() {
        activity.removeEmptyInfiniteLoop();
    }

    public String toXML() {
        String s = new String("<TASK name=\"" + name + "\" priority=\"" + priority + "\" >\n");
        for (TMLAttribute attr : attributes) {
            s += attr.toXML();
        }
        s += activity.toXML();
        s += "</TASK>\n";
        return s;
    }


    // returns -1 if the WC cannot be computed
    // The function follows one path randomly and concatenate ExecI operators
    public int getWorstCaseIComplexity() {
        if (activity == null) {
            return -1;
        }
        //TraceManager.addDev("Handling task:" + getTaskName());
        return activity.getWorstCaseIComplexity();
    }

    public boolean equalSpec(Object o) {
        if (!(o instanceof TMLTask)) return false;
        if (!super.equalSpec(o)) return false;
        TMLTask tmlTask = (TMLTask) o;
        if(!(new HashSet<>(attributes).equals(new HashSet<>(tmlTask.attributes))))
            return false;
        return operationType == tmlTask.operationType &&
                isDaemon == tmlTask.isDaemon &&
                isAttacker == tmlTask.isAttacker &&
                operation.equals(tmlTask.operation) &&
                operationMEC.equals(tmlTask.operationMEC);
    }
}
